### PR TITLE
Add preliminary permissions module

### PR DESCRIPTION
**Pull request by danlamanna**
This adds a VERY WIP version of permissions.

### CODE REVIEW

File: isic/ingest/models/contributor.py [hunk 1]
```diff
@@ -49,3 +49,33 @@ class Contributor(CreationSortedTimeStampedModel):
 
     def __str__(self) -> str:
         return self.institution_name
+
+
+class ContributorPermissions:
+    model = Contributor
```
[the remark this answers is not in the document]
Comment (danlamanna) [reply isic/ingest/models/contributor.py:55]: Which of these approaches do you prefer? / How critical is it to fixing this api before we merge the permissions?

File: isic/ingest/models/contributor.py [hunk 2]
```diff
@@ -49,3 +49,33 @@ class Contributor(CreationSortedTimeStampedModel):
...
+
+        return qs.filter(owners__in=[user_obj])
+
+    @staticmethod
+    def view_contributor(user_obj, obj):
+        # TODO: use .contains in django 4
+        return ContributorPermissions.view_contributor_list(user_obj).filter(pk=obj.pk).exists()
+
+    @staticmethod
+    def add_contributor(user_obj, obj=None):
```
Comment (brianhelba) [inline isic/ingest/models/contributor.py:77]: Maybe this should be attached to `User`, not `Contributor`, to make the argument not be a different (or non-existant) type.

File: isic/ingest/views/upload.py [hunk 3]
```diff
@@ -4,33 +4,25 @@
...
-    ctx = {'contributors': contributors}
+    ctx = {'contributors': get_visible_objects(request.user, 'ingest.view_contributor')}
     if ctx['contributors'].count() == 0:
         return HttpResponseRedirect(reverse('upload/create-contributor'))
 
     return render(request, 'ingest/contributor_select_or_create.html', ctx)
 
 
-@login_required
+@permission_or_404('ingest.view_contributor', (Contributor, 'pk', 'contributor_pk'))
```
Comment (brianhelba) [inline isic/ingest/views/upload.py:21]: Maybe this should be a separate `create_cohort` permission.

File: isic/ingest/views/metadata.py [hunk 4]
```diff
@@ -34,20 +34,19 @@ def __init__(self, user, cohort, *args, **kwargs):
...
+                for m in get_visible_objects(
+                    user, 'ingest.view_metadatafile', MetadataFile.objects.filter(cohort=cohort)
                 )
             ],
             widget=forms.RadioSelect,
         )
 
 
 @login_required
+@permission_or_404('ingest.view_cohort', (Cohort, 'pk', 'cohort_pk'))
```
Comment (brianhelba) [inline isic/ingest/views/metadata.py:46]: Separate `create_zip_file` permission?

File: isic/core/views.py [hunk 5]
```diff
@@ -97,6 +100,8 @@ def collection_detail(request, pk):
     )
 
 
+# TODO: refactor permissions in the future to use permission_or_404 and filtering
```
Comment (brianhelba) [inline isic/core/views.py:103]: TODO

File: isic/core/views.py [hunk 6]
```diff
@@ -54,16 +55,18 @@ def stats(request):
...
 def staff_list(request):
     users = User.objects.filter(is_staff=True).order_by('email')
     return render(request, 'core/staff_list.html', {'users': users, 'total_users': User.objects})
 
 
-@staff_member_required
 def collection_list(request):
-    collections = Collection.objects.annotate(num_images=Count('images', distinct=True)).order_by(
-        '-name'
+    # TODO: should the image count be access controlled too?
```
Comment (brianhelba) [inline isic/core/views.py:65]: TODO

**Review by brianhelba — APPROVED** [review]

**Pull request merged**